core/types/calls.cc

```diff
@@ -2395,20 +2395,43 @@ class Array_flatten : public IntrinsicMethod {
             // This only shows up because t->elementType(gs) for tuples returns an OrType of all its elements.
             // So to properly handle nested tuples, we have to descend into the OrType's.
             [&](const OrType &o) {
-                result = Types::any(gs, recursivelyFlattenArrays(gs, o.left, newDepth),
-                                    recursivelyFlattenArrays(gs, o.right, newDepth));
+                result = Types::any(gs, recursivelyFlattenArrays(gs, args, o.left, newDepth),
+                                    recursivelyFlattenArrays(gs, args, o.right, newDepth));
+            },
+
+            [&](const ClassType &c) {
+                // If the element type supports the #to_ary method, then Ruby will implicitly call it when
+                // flattening. So here we need to get the return type of calling #to_ary on an element.
+                auto toAry = c.symbol.data(gs)->findMember(gs, Names::toAry());
+                if (toAry.exists()) {
```

Comment:
We can unconditionally dispatch here, and then discard the result if there are errors.  See the `Magic_splat` implementation in this same file for how to do that.